src/Plugin/authentication/OAuth2ServerAuthentication.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace Drupal\restful\Plugin\authentication;
+
+use Drupal\Component\Plugin\PluginBase;
+use Drupal\restful\Exception\UnauthorizedException;
+use Drupal\restful\Http\RequestInterface;
+use Drupal\restful\Plugin\ResourcePluginManager;
+
+/**
+ * Authentication support for oauth2_server.
+ *
+ * @Authentication(
+ *   id = "oauth2_auth",
+ *   label = "OAuth2 authentication",
+ *   description = "Authenticate requests based on oauth2_server auth.",
+ * )
+ */
+class OAuth2ServerAuthentication extends Authentication {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applies(RequestInterface $request) {
+    return module_exists('oauth2_server') && $this->getResourcePluginIdFromRequest();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function authenticate(RequestInterface $request) {
+    $oauth2_info = $this->getOAuth2Info($request);
+    if (!$oauth2_info) {
+      return NULL;
+    }
+
+    $result = oauth2_server_check_access($oauth2_info['server'], $oauth2_info['scope']);
+    if ($result instanceof \OAuth2\Response) {
+      throw new UnauthorizedException($result->getResponseBody(), $result->getStatusCode());
+    }
+    elseif (is_array($result) && !empty($result['user_id'])) {
```

Comment:
For clarity, I meant «Make this whole `elseif`:»

I'd leave the `if` above alone, but simplify the logic (without leaving any execution path without a `return`).